joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -161,7 +203,9 @@ def configure_robot(self, description):
         self.free_joints = {}
         self.joint_list = [] # for maintaining the original order of the joints
 
-        if robot.getElementsByTagName('COLLADA'):
+        if robot.getElementsByTagName('sdf'):
+            self.init_sdf(robot)
+        elif robot.getElementsByTagName('COLLADA'):
```

Comment:
I know you are just extending this, but this check seems too broad.  That is, if I understand https://docs.python.org/3.6/library/xml.dom.html#xml.dom.Document.getElementsByTagName correctly, this will look in the entire XML tree for tags of this name.  I think this has some (limited) potential to produce false positives.  If someone was using COLLADA, but named one of their joints 'sdf', I think we would attempt to parse it as SDF.  Is there something else we can check in the document that is more specific?  The root element name, for instance?

Reply:
Yeah, it seems too broad to me too. I think checking the root element would be a better solution. The C++ parser is handling it a little different by scoring based on the location of the first tag text, like `<COLLADA`

https://github.com/ros/collada_urdf/pull/39/files#diff-f13e149a001d68baae11ac5da7f08dc5R56-R59

but that would be fooled by the text being in a comment in the xml.

```xml
<!-- this is not a <COLLADA> file -->
<sdf>
...
```

It would be nice if joint_state_publisher used the C++ parser so this would only have to be fixed in one place.

Reply:
> It would be nice if joint_state_publisher used the C++ parser so this would only have to be fixed in one place.

While I somewhat agree, I also don't relish the thought of putting a CPython binding on top of the URDF ecosystem.  At the very least, it means we don't have a universal wheel for Python anymore.

Reply:
> I also don't relish the thought of putting a CPython binding on top of the URDF ecosystem. At the very least, it means we don't have a universal wheel for Python anymore.

Where were you thinking about the CPython binding living? `urdf_parser_py`? I was thinking of `joint_state_publisher` having a CPython extension that calls the urdf C++ api to parse the robot description and return an object with the joint info joint_state_publisher needs.

Reply:
> Where were you thinking about the CPython binding living? `urdf_parser_py`? I was thinking of `joint_state_publisher` having a CPython extension that calls the urdf C++ api to parse the robot description and return an object with the joint info joint_state_publisher needs.

I didn't have a specific place in mind.

But even in the case you are suggesting, its not clear to me that providing a CPython extension for joint_state_publisher is a win.  It seems to me the pros and cons of switching to a CPython extension:

Pros:
* Reduce duplication in parsing the XML.
* Get support for new formats for "free" once they are in urdf_parser

Cons:
* Replace Python code most Python programmers can understand with more complex CPython binding.
* joint_state_publisher is no longer a [Pure python](https://packaging.python.org/guides/distributing-packages-using-setuptools/#pure-python-wheels) wheel.

Reply:
It's been a few years. Looking back, reusing the existing URDF/SDformat parsers instead of partially implementing parsers in this package seems like a big advantage, and using another language we're already extensively using in ROS seems like a small disadvantage.

I'm not about the pure python wheel con. It looks like [`colcon` doesn't build packages as wheels](https://github.com/colcon/colcon-core/blob/99c03262ad7edd4af5954c424c9248e0c4fe7307/colcon_core/task/python/build.py#L102-L104), and if someone were building the package without using colcon I'm not sure what they would find more advantageous.